semgrep/semgrep/core_runner.py

```diff
@@ -99,6 +101,140 @@ def uniq_error_id(
     return (error.code, error.path, error.start, error.end, error.message)
 
 
+class StreamingSemgrepCore:
+    """
+    Handles running semgrep-core in a streaming fashion
+
+    This behavior is assumed to be that semgrep-core:
+    - prints a "." on a newline for every file it finishes scanning
+    - prints a single json blob of all results
+
+    Exposes the subprocess.CompletedProcess properties for
+    expediency in integrating
+    """
+
+    def __init__(self, cmd: List[str], total: int) -> None:
+        """
+        cmd: semgrep-core command to run
+        total: how many rules to run / how many "." we expect to see
+               used to display progress_bar
+        """
+        self._cmd = cmd
+        self._total = total
+        self._stdout = ""
+        self._stderr = ""
+        self._progress_bar: Optional[tqdm] = None  # type: ignore
+
+    @property
+    def stdout(self) -> str:
+        # stdout of semgrep-core sans "."
+        return self._stdout
+
+    @property
+    def stderr(self) -> str:
+        # stderr of semgrep-core command
+        return self._stderr
+
+    async def _core_stdout_processor(
+        self, stream: Optional[asyncio.StreamReader]
+    ) -> None:
+        """
+        Asynchronously process stdout of semgrep-core
+
+        Updates progress bar one increment for every "." it sees from semgrep-core
+        stdout
+
+        When it sees non-"." output it saves it to self._stdout
+        """
+        # appease mypy. stream is only None if call to create_subproccess_exec
+        # sets stdout/stderr stream to None
+        assert stream
+        while True:
+            # blocking read if buffer doesnt contain any lines or EOF
+            line_bytes = await stream.readline()
+
+            # readline returns None on EOF
+            if not line_bytes:
```

Comment:
```suggestion
            if line_bytes is None:
```
I assume in some situation it could return `b""` and we don't wanna break then? or if we update this code to do some preprocessing on the line, like `line_bytes.strip()` or something

Reply:
I think it has to return None to mark EOF otherwise it was intentionally an empty line

Reply:
Ah yes you're right

Reply:
Double checked and docs say it returns empty buffer on EOF https://docs.python.org/3/library/asyncio-stream.html#asyncio.StreamReader.readline